### DISCUSSION TITLE
Capturing HTTPS Request in Transparant mode without certificate

### DISCUSSION THREAD

**Discussion by DeBroin** [mitmproxy/mitmproxy#5641] [Q&A] [answered]
Hello mitm guru's!

I've spent the last few weeks on an issue I can't seem to resolve on my own.
I'm hoping someone here will see this and be able to provide help, clues or insight.

My problem:
I have an device on my network that sends Https data. I have very limited control over the device (in the sense that there aren't options to trust a certificate, use a proxy, etc). However, I would like to see the traffic it sends, and be able to read it (as in: needs to be decrypted).

I don't care if the request gets dropped after decrypting, I also don't care about any response the server sends back. I just need to be able to read what gets sent.

What I have done:
- I set up mitmproxy in transparent mode on a Linux VM (internal IP 192.168.0.204).
- On that Linux VM I did the following:
   - Set up mitmproxy in transparent mode according to [this link](https://docs.mitmproxy.org/stable/howto-transparent/) in the official docs
   - Trusted the mitmproxy certificate [via this link](https://askubuntu.com/questions/73287/how-do-i-install-a-root-certificate/94861#94861), which is a link in the official mitm doc (don't know if this was needed, can anyone confirm? Does mitmproxy in transparent mode generate a certificate on the fly for each request, without needing to have it trusted on the machine that runs mitm itself?)
- I have set up a windows machine with a fixed IP (192.168.0.128) that has the IP of the Linux VM as gateway (IP 192.168.0.204).
- My windows machine also has a trusted mitm certificate (please bear in mind that I will not be able to trust the certificate on the actual device that I want to capture, I just use a Windows machine at the moment until things are set up properly)

Once I set up the default gateway, I can see traffic coming through mitmproxy, but I'm getting a lot of TLS Handshake failures, pretty much for everything iirc.
I messed around with setting the proxy on the linux machine to itself (by setting a manual proxy in OS settings - Network), but I seem to be going around in circles, and I'm hitting walls no matter what I try.

TL-DR version:
I have a device on my network for which I cannot set a certificate (only a gateway), I want to inspect the calls it sends. I do not care about a response or proper handshake etc. Just the decrypted request.

Can anybody please shed some light on what I'm missing or doing wrong?

**Comment by mhils** [comment, marked as answer]
In short, your client (the one at the end - the device on your network) needs to trust the proxy's certificate. Otherwise mitmproxy (or any other TLS-intercepting proxy) will not be able read _any_ message contents (not even requests).

**Reply by DeBroin** [reply]
Is there really, absolutely, no other way but to have the client trust the certificate?
Would I be able to do something with mitmdump, tcpdump or something similar, along with SSLKEYLOGFILE?
I am able to intercept the data via wireshark, so I'm not sure if that could be a solution?
Apologies if these are silly questions/remarks, but I'm not overly familiar with networking stuff.

Many thanks in advance!

**Reply by mhils** [reply]
> Is there really, absolutely, no other way but to have the client trust the certificate?

If the client does not trust the certificate (and there are no vulnerabilities you can exploit[^1]), you can only observe metadata, i.e. which endpoints your device is communicating with. You cannot observe encrypted message contents.


[^1]: This hasn't been practically relevant for years.

**Reply by DeBroin** [reply]
Well... Not the answer I was hoping for, but I'm really glad I got a definitive yes/no, and in such a short amount of time. Thanks, you saved me a few days of headaches pursuing this :)

**Reply by RHAD1969** [reply]
> > Is there really, absolutely, no other way but to have the client trust the certificate?
> 
> If the client does not trust the certificate (and there are no vulnerabilities you can exploit[1](#user-content-fn-1-cd0b02c66443c75e803430071773c120)), you can only observe metadata, i.e. which endpoints your device is communicating with. You cannot observe encrypted message contents.
> 
> ## Footnotes
> 1. This hasn't been practically relevant for years. [↩](#user-content-fnref-1-cd0b02c66443c75e803430071773c120)

This is precisely what I want to do: just capture the encrypted data, no need to decrypt them. I'm only interested in the destination. I don't need the encrypted content, but keep facing the "Client TLS handshake failed" error.

How can I capture all raw data passing my mitmproxy?

**Reply by mhils** [reply]
If you don't decrypt anything at all you may want to use something like tcpdump instead of mitmproxy, but the mitmproxy answer is the ignore domains feature: https://docs.mitmproxy.org/stable/howto-ignoredomains/
